server/lib/matchmaking/matchmaking-api.ts

```diff
@@ -0,0 +1,124 @@
+import Router, { RouterContext } from '@koa/router'
+import httpErrors from 'http-errors'
+import Joi from 'joi'
+import Koa from 'koa'
+import { container } from 'tsyringe'
+import { assertUnreachable } from '../../../common/assert-unreachable'
+import { ALL_MATCHMAKING_TYPES } from '../../../common/matchmaking'
+import { httpApi, HttpApi } from '../http/http-api'
+import ensureLoggedIn from '../session/ensure-logged-in'
+import createThrottle from '../throttle/create-throttle'
+import throttleMiddleware from '../throttle/middleware'
+import { validateRequest } from '../validation/joi-validator'
+import {
+  MatchmakingService,
+  MatchmakingServiceError,
+  MatchmakingServiceErrorCode,
+} from './matchmaking-service'
+
+const matchmakingThrottle = createThrottle('matchmaking', {
+  rate: 20,
+  burst: 40,
+  window: 60000,
+})
+
+function convertMatchmakingServiceError(err: Error) {
+  if (!(err instanceof MatchmakingServiceError)) {
+    throw err
+  }
+
+  switch (err.code) {
+    case MatchmakingServiceErrorCode.UserOffline:
+      throw new httpErrors.NotFound(err.message)
+    case MatchmakingServiceErrorCode.InvalidMapPool:
+    case MatchmakingServiceErrorCode.InvalidMaps:
+    case MatchmakingServiceErrorCode.ClientDisconnected:
+    case MatchmakingServiceErrorCode.NotInQueue:
+    case MatchmakingServiceErrorCode.NoActiveMatch:
+      throw new httpErrors.BadRequest(err.message)
+    case MatchmakingServiceErrorCode.MatchmakingDisabled:
+      throw new httpErrors.Forbidden(err.message)
+    case MatchmakingServiceErrorCode.GameplayConflict:
+      throw new httpErrors.Conflict(err.message)
+    default:
+      assertUnreachable(err.code)
+  }
+}
+
+async function convertMatchmakingServiceErrors(ctx: RouterContext, next: Koa.Next) {
+  try {
+    await next()
+  } catch (err) {
+    convertMatchmakingServiceError(err)
+  }
+}
+
+@httpApi()
+export class MatchmakingApi extends HttpApi {
+  constructor() {
+    super('/matchmaking')
+    container.resolve(MatchmakingService)
+  }
+
+  applyRoutes(router: Router): void {
+    router
+      .use(ensureLoggedIn, convertMatchmakingServiceErrors)
+      .post(
+        '/find',
+        throttleMiddleware(matchmakingThrottle, ctx => String(ctx.session!.userId)),
+        findMatch,
+      )
+      .delete(
+        '/',
+        throttleMiddleware(matchmakingThrottle, ctx => String(ctx.session!.userId)),
+        cancelSearch,
+      )
+      .post(
+        '/',
+        throttleMiddleware(matchmakingThrottle, ctx => String(ctx.session!.userId)),
+        acceptMatch,
+      )
+  }
+}
+
+async function findMatch(ctx: RouterContext) {
```

Comment:
Don't have to fix this in this PR, but I'd *really* recommend moving to the API helper I wrote (so you don't have to call `validateRequest` manually and the response body is properly typed) + probably putting these methods on the `HttpApi` object (so you don't have to inject in every method)

Reply:
Gotta be honest here and mention that I actually prefer the syntax of explicitly calling the `validateRequest` in the handler body, instead of sending it as a parameter to a helper function. Purely from an aesthetic/readability perspective.

I like the benefit of using it to handle the API response though.